teamcity: failed tests on master: test/TestLogic, test/TestLogic/5node, test/TestLogic/5node/distsql_agg, acceptance/TestAdminLossOfQuorum

The following tests appear to have failed:

[#266803](https://teamcity.cockroachdb.com/viewLog.html?buildId=266803):

```
--- FAIL: test/TestLogic (26.020s)
test_log_scope.go:80: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic497036230
test_log_scope.go:63: use -show-logs to present logs inline
--- FAIL: test/TestLogic/5node (0.000s)
null
--- FAIL: test/TestLogic/5node/distsql_agg (3.180s)
logic_test.go:1718: 
	testdata/logic_test/distsql_agg:373: expected success, but found
	pq: snapshot failed: (n1,s1):?: remote couldn't accept snapshot with error: [n1,s1],r21: cannot apply snapshot: snapshot intersects existing range [n1,s1,r10/1:/{Table/51/1/9-Max}]
logic_test.go:1728: 
	testdata/logic_test/distsql_agg:373: error in statement, skipping to next file
--- FAIL: acceptance/TestAdminLossOfQuorum (10.130s)
test_log_scope.go:80: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestAdminLossOfQuorum369153985
test_log_scope.go:63: use -show-logs to present logs inline
admin_test.go:71: Get https://127.0.0.1:32967/_status/nodes: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
------- Stdout: -------
xenial-20170214: Pulling from library/ubuntu
Digest: sha256:dd7808d8792c9841d0b460122f1acf0a2dd1f56404f8d1e56298048885e45535
Status: Image is up to date for ubuntu:xenial-20170214
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestAdminLossOfQuorum369153985

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/sql/expand_plan.go
+++ b/pkg/sql/expand_plan.go
@@ -223,7 +223,7 @@ func doExpandPlan(
 	case *splitNode:
 		n.rows, err = doExpandPlan(ctx, p, noParams, n.rows)
 
-	case *relocateNode:
+	case *testingRelocateNode:
 		n.rows, err = doExpandPlan(ctx, p, noParams, n.rows)
 
 	case *valuesNode:
@@ -535,7 +535,7 @@ func simplifyOrderings(plan planNode, usefulOrdering sqlbase.ColumnOrdering) pla
 	case *splitNode:
 		n.rows = simplifyOrderings(n.rows, nil)
 
-	case *relocateNode:
+	case *testingRelocateNode:
 		n.rows = simplifyOrderings(n.rows, nil)
 
 	case *valuesNode:

--- a/pkg/sql/filter_opt.go
+++ b/pkg/sql/filter_opt.go
@@ -314,7 +314,7 @@ func (p *planner) propagateFilters(
 			return plan, extraFilter, err
 		}
 
-	case *relocateNode:
+	case *testingRelocateNode:
 		if n.rows, err = p.triggerFilterPropagation(ctx, n.rows); err != nil {
 			return plan, extraFilter, err
 		}

--- a/pkg/sql/limit_opt.go
+++ b/pkg/sql/limit_opt.go
@@ -163,7 +163,7 @@ func applyLimit(plan planNode, numRows int64, soft bool) {
 	case *splitNode:
 		setUnlimited(n.rows)
 
-	case *relocateNode:
+	case *testingRelocateNode:
 		setUnlimited(n.rows)
 
 	case *valuesNode:

--- a/pkg/sql/needed_columns.go
+++ b/pkg/sql/needed_columns.go
@@ -185,7 +185,7 @@ func setNeededColumns(plan planNode, needed []bool) {
 	case *splitNode:
 		setNeededColumns(n.rows, allColumns(n.rows))
 
-	case *relocateNode:
+	case *testingRelocateNode:
 		setNeededColumns(n.rows, allColumns(n.rows))
 
 	case *alterTableNode:

--- a/pkg/sql/parser/split.go
+++ b/pkg/sql/parser/split.go
@@ -42,9 +42,9 @@ func (node *Split) Format(buf *bytes.Buffer, f FmtFlags) {
 	FormatNode(buf, f, node.Rows)
 }
 
-// Relocate represents an `ALTER TABLE/INDEX .. TESTING_RELOCATE ..`
+// TestingRelocate represents an `ALTER TABLE/INDEX .. TESTING_RELOCATE ..`
 // statement.
-type Relocate struct {
+type TestingRelocate struct {
 	// Only one of Table and Index can be set.
 	Table *NormalizableTableName
 	Index *TableNameWithIndex
@@ -55,7 +55,7 @@ type Relocate struct {
 }
 
 // Format implements the NodeFormatter interface.
-func (node *Relocate) Format(buf *bytes.Buffer, f FmtFlags) {
+func (node *TestingRelocate) Format(buf *bytes.Buffer, f FmtFlags) {
 	buf.WriteString("ALTER ")
 	if node.Index != nil {
 		buf.WriteString("INDEX ")

--- a/pkg/sql/parser/sql.go
+++ b/pkg/sql/parser/sql.go
@@ -7392,14 +7392,14 @@ sqldefault:
 		//line sql.y:1847
 		{
 			/* SKIP DOC */
-			sqlVAL.union.val = &Relocate{Table: sqlDollar[3].union.newNormalizableTableName(), Rows: sqlDollar[5].union.slct()}
+			sqlVAL.union.val = &TestingRelocate{Table: sqlDollar[3].union.newNormalizableTableName(), Rows: sqlDollar[5].union.slct()}
 		}
 	case 263:
 		sqlDollar = sqlS[sqlpt-5 : sqlpt+1]
 		//line sql.y:1851
 		{
 			/* SKIP DOC */
-			sqlVAL.union.val = &Relocate{Index: sqlDollar[3].union.tableWithIdx(), Rows: sqlDollar[5].union.slct()}
+			sqlVAL.union.val = &TestingRelocate{Index: sqlDollar[3].union.tableWithIdx(), Rows: sqlDollar[5].union.slct()}
 		}
 	case 264:
 		sqlDollar = sqlS[sqlpt-4 : sqlpt+1]

--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -1867,12 +1867,12 @@ testing_relocate_stmt:
   ALTER TABLE qualified_name TESTING_RELOCATE select_stmt
   {
     /* SKIP DOC */
-    $$.val = &Relocate{Table: $3.newNormalizableTableName(), Rows: $5.slct()}
+    $$.val = &TestingRelocate{Table: $3.newNormalizableTableName(), Rows: $5.slct()}
   }
 | ALTER INDEX table_name_with_index TESTING_RELOCATE select_stmt
   {
     /* SKIP DOC */
-    $$.val = &Relocate{Index: $3.tableWithIdx(), Rows: $5.slct()}
+    $$.val = &TestingRelocate{Index: $3.tableWithIdx(), Rows: $5.slct()}
   }
 
 scatter_stmt:

--- a/pkg/sql/parser/stmt.go
+++ b/pkg/sql/parser/stmt.go
@@ -310,10 +310,10 @@ func (n *RenameTable) StatementTag() string {
 }
 
 // StatementType implements the Statement interface.
-func (*Relocate) StatementType() StatementType { return Rows }
+func (*TestingRelocate) StatementType() StatementType { return Rows }
 
 // StatementTag returns a short string identifying the type of statement.
-func (*Relocate) StatementTag() string { return "TESTING_RELOCATE" }
+func (*TestingRelocate) StatementTag() string { return "TESTING_RELOCATE" }
 
 // StatementType implements the Statement interface.
 func (*Restore) StatementType() StatementType { return Rows }
@@ -632,7 +632,7 @@ func (n *ParenSelect) String() string              { return AsString(n) }
 func (n *PauseJob) String() string                 { return AsString(n) }
 func (n *Prepare) String() string                  { return AsString(n) }
 func (n *ReleaseSavepoint) String() string         { return AsString(n) }
-func (n *Relocate) String() string                 { return AsString(n) }
+func (n *TestingRelocate) String() string          { return AsString(n) }
 func (n *RenameColumn) String() string             { return AsString(n) }
 func (n *RenameDatabase) String() string           { return AsString(n) }
 func (n *RenameIndex) String() string              { return AsString(n) }

--- a/pkg/sql/plan.go
+++ b/pkg/sql/plan.go
@@ -154,7 +154,7 @@ var _ planNode = &insertNode{}
 var _ planNode = &joinNode{}
 var _ planNode = &limitNode{}
 var _ planNode = &ordinalityNode{}
-var _ planNode = &relocateNode{}
+var _ planNode = &testingRelocateNode{}
 var _ planNode = &renderNode{}
 var _ planNode = &scanNode{}
 var _ planNode = &scatterNode{}
@@ -358,8 +358,8 @@ func (p *planner) newPlan(
 		return p.newPlan(ctx, n.Select, desiredTypes)
 	case *parser.PauseJob:
 		return p.PauseJob(ctx, n)
-	case *parser.Relocate:
-		return p.Relocate(ctx, n)
+	case *parser.TestingRelocate:
+		return p.TestingRelocate(ctx, n)
 	case *parser.RenameColumn:
 		return p.RenameColumn(ctx, n)
 	case *parser.RenameDatabase:
@@ -489,8 +489,8 @@ func (p *planner) prepare(ctx context.Context, stmt parser.Statement) (planNode,
 		return p.ShowRanges(ctx, n)
 	case *parser.Split:
 		return p.Split(ctx, n)
-	case *parser.Relocate:
-		return p.Relocate(ctx, n)
+	case *parser.TestingRelocate:
+		return p.TestingRelocate(ctx, n)
 	case *parser.Scatter:
 		return p.Scatter(ctx, n)
 	case *parser.Update:

--- a/pkg/sql/plan_columns.go
+++ b/pkg/sql/plan_columns.go
@@ -82,7 +82,7 @@ func getPlanColumns(plan planNode, mut bool) sqlbase.ResultColumns {
 		// Nodes with a fixed schema.
 	case *explainDistSQLNode:
 		return n.getColumns(mut, explainDistSQLColumns)
-	case *relocateNode:
+	case *testingRelocateNode:
 		return n.getColumns(mut, relocateNodeColumns)
 	case *scatterNode:
 		return n.getColumns(mut, scatterNodeColumns)

--- a/pkg/sql/split_at.go
+++ b/pkg/sql/split_at.go
@@ -157,10 +157,12 @@ var splitNodeColumns = sqlbase.ResultColumns{
 	},
 }
 
-// Relocate moves ranges to specific stores
+// TestingRelocate moves ranges to specific stores
 // (`ALTER TABLE/INDEX ... TESTING_RELOCATE ...` statement)
 // Privileges: INSERT on table.
-func (p *planner) Relocate(ctx context.Context, n *parser.Relocate) (planNode, error) {
+func (p *planner) TestingRelocate(
+	ctx context.Context, n *parser.TestingRelocate,
+) (planNode, error) {
 	tableDesc, index, err := p.getTableAndIndex(ctx, n.Table, n.Index, privilege.INSERT)
 	if err != nil {
 		return nil, err
@@ -206,7 +208,7 @@ func (p *planner) Relocate(ctx context.Context, n *parser.Relocate) (planNode, e
 		}
 	}
 
-	return &relocateNode{
+	return &testingRelocateNode{
 		p:         p,
 		tableDesc: tableDesc,
 		index:     index,
@@ -215,7 +217,7 @@ func (p *planner) Relocate(ctx context.Context, n *parser.Relocate) (planNode, e
 	}, nil
 }
 
-type relocateNode struct {
+type testingRelocateNode struct {
 	optColumnsSlot
 
 	p                 *planner
@@ -229,7 +231,7 @@ type relocateNode struct {
 	storeMap map[roachpb.StoreID]roachpb.NodeID
 }
 
-func (n *relocateNode) Start(ctx context.Context) error {
+func (n *testingRelocateNode) Start(ctx context.Context) error {
 	return n.rows.Start(ctx)
 }
 
@@ -255,7 +257,7 @@ func lookupRangeDescriptor(
 	return desc, nil
 }
 
-func (n *relocateNode) Next(ctx context.Context) (bool, error) {
+func (n *testingRelocateNode) Next(ctx context.Context) (bool, error) {
 	// Each Next call relocates one range (corresponding to one row from n.rows).
 	// TODO(radu): perform multiple relocations in parallel.
 
@@ -311,21 +313,21 @@ func (n *relocateNode) Next(ctx context.Context) (bool, error) {
 	}
 	n.lastRangeStartKey = rangeDesc.StartKey.AsRawKey()
 
-	if err := storage.RelocateRange(ctx, n.p.ExecCfg().DB, rangeDesc, targets); err != nil {
+	if err := storage.TestingRelocateRange(ctx, n.p.ExecCfg().DB, rangeDesc, targets); err != nil {
 		return false, err
 	}
 
 	return true, nil
 }
 
-func (n *relocateNode) Values() parser.Datums {
+func (n *testingRelocateNode) Values() parser.Datums {
 	return parser.Datums{
 		parser.NewDBytes(parser.DBytes(n.lastRangeStartKey)),
 		parser.NewDString(keys.PrettyPrint(n.lastRangeStartKey)),
 	}
 }
 
-func (n *relocateNode) Close(ctx context.Context) {
+func (n *testingRelocateNode) Close(ctx context.Context) {
 	n.rows.Close(ctx)
 }
 

--- a/pkg/sql/walk.go
+++ b/pkg/sql/walk.go
@@ -273,7 +273,7 @@ func (v *planVisitor) visit(plan planNode) {
 	case *splitNode:
 		v.visit(n.rows)
 
-	case *relocateNode:
+	case *testingRelocateNode:
 		v.visit(n.rows)
 
 	case *insertNode:
@@ -512,7 +512,7 @@ var planNodeNames = map[reflect.Type]string{
 	reflect.TypeOf(&joinNode{}):             "join",
 	reflect.TypeOf(&limitNode{}):            "limit",
 	reflect.TypeOf(&ordinalityNode{}):       "ordinality",
-	reflect.TypeOf(&relocateNode{}):         "relocate",
+	reflect.TypeOf(&testingRelocateNode{}):  "testingRelocate",
 	reflect.TypeOf(&renderNode{}):           "render",
 	reflect.TypeOf(&scanNode{}):             "scan",
 	reflect.TypeOf(&scatterNode{}):          "scatter",

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -28,6 +28,7 @@ import (
 	"math"
 	"math/rand"
 	"reflect"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -3787,12 +3788,12 @@ func evalLeaseInfo(
 	return EvalResult{}, nil
 }
 
-// RelocateRange relocates a given range to a given set of stores. The first
+// TestingRelocateRange relocates a given range to a given set of stores. The first
 // store in the slice becomes the new leaseholder.
 //
 // This is best-effort; if replication queues are enabled and a change in
 // membership happens at the same time, there will be errors.
-func RelocateRange(
+func TestingRelocateRange(
 	ctx context.Context,
 	db *client.DB,
 	rangeDesc roachpb.RangeDescriptor,
@@ -3817,22 +3818,50 @@ func RelocateRange(
 			addTargets = append(addTargets, t)
 		}
 	}
-	if len(addTargets) > 0 {
-		if err := db.AdminChangeReplicas(
-			ctx, rangeDesc.StartKey.AsRawKey(), roachpb.ADD_REPLICA, addTargets,
-		); err != nil {
+
+	canRetry := func(err error) bool {
+		whitelist := []string{
+			"snapshot intersects existing range",
+		}
+		for _, substr := range whitelist {
+			if strings.Contains(err.Error(), substr) {
+				return true
+			}
+		}
+		return false
+	}
+
+	for len(addTargets) > 0 {
+		if err := ctx.Err(); err != nil {
 			return err
 		}
+
+		target := addTargets[0]
+		if err := db.AdminChangeReplicas(
+			ctx, rangeDesc.StartKey.AsRawKey(), roachpb.ADD_REPLICA, []roachpb.ReplicationTarget{target},
+		); err != nil {
+			returnErr := errors.Wrapf(err, "while adding target %v", target)
+			if !canRetry(err) {
+				return returnErr
+			}
+			log.Warning(ctx, returnErr)
+			continue
+		}
+		addTargets = addTargets[1:]
 	}
 
 	// Step 2: Transfer the lease to the first target. This needs to happen before
 	// we remove replicas or we may try to remove the lease holder.
 
-	if err := db.AdminTransferLease(
-		ctx, rangeDesc.StartKey.AsRawKey(), targets[0].StoreID,
-	); err != nil {
-		return err
+	transferLease := func() {
+		if err := db.AdminTransferLease(
+			ctx, rangeDesc.StartKey.AsRawKey(), targets[0].StoreID,
+		); err != nil {
+			log.Warningf(ctx, "while transferring lease: %s", err)
+		}
 	}
+
+	transferLease()
 
 	// Step 3: Remove any replicas that are not targets.
 
@@ -3852,14 +3881,26 @@ func RelocateRange(
 			})
 		}
 	}
-	if len(removeTargets) > 0 {
-		if err := db.AdminChangeReplicas(
-			ctx, rangeDesc.StartKey.AsRawKey(), roachpb.REMOVE_REPLICA, removeTargets,
-		); err != nil {
+
+	for len(removeTargets) > 0 {
+		if err := ctx.Err(); err != nil {
 			return err
 		}
+
+		target := removeTargets[0]
+		transferLease()
+		if err := db.AdminChangeReplicas(
+			ctx, rangeDesc.StartKey.AsRawKey(), roachpb.REMOVE_REPLICA, []roachpb.ReplicationTarget{target},
+		); err != nil {
+			log.Warningf(ctx, "while removing target %v: %s", target, err)
+			if !canRetry(err) {
+				return err
+			}
+			continue
+		}
+		removeTargets = removeTargets[1:]
 	}
-	return nil
+	return ctx.Err()
 }
 
 // adminScatter moves replicas and leaseholders for a selection of ranges.
@@ -3893,7 +3934,7 @@ func (r *Replica) adminScatter(
 	}
 	targets := stores[:num]
 
-	relocateErr := RelocateRange(ctx, db, rangeDesc, targets)
+	relocateErr := TestingRelocateRange(ctx, db, rangeDesc, targets)
 
 	res := roachpb.AdminScatterResponse{
 		Ranges: []roachpb.AdminScatterResponse_Range{{

